File: 36/valid_sudoku.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cmath>

using namespace std;



bool isValidSudoku(const vector<vector<char>> &board) {

	int length = board.size();

	vector<set<char>> rows(length), cols(length), sub_boards(length);
	pair<set<char>::iterator, bool> hasValue;

	for(size_t i{0}; i < length; ++i) {
		for(size_t j{0}; j < board[0].size(); ++j) {
			int sub_board_idx = 3*floor(i/3) + floor(j/3);

			char cell = board[i][j];
			if (cell == '.') continue;

			hasValue = rows[i].insert(cell);
			if (!hasValue.second) return false;

			hasValue = cols[j].insert(cell);
			if (!hasValue.second) return false;

			hasValue = sub_boards[sub_board_idx].insert(cell);
			if (!hasValue.second) return false;
		}
	}

	return true;
}

int main() {


	const vector<vector<char>> board  = {
		{'5','3','.','.','7','.','.','.','.'},
		{'6','.','.','1','9','5','.','.','.'},
		{'.','9','8','.','.','.','.','6','.'},
		{'8','.','.','.','6','.','.','.','3'},
		{'4','.','.','8','.','3','.','.','1'},
		{'7','.','.','.','2','.','.','.','6'},
		{'.','6','.','.','.','.','2','8','.'},
		{'.','.','.','4','1','9','.','.','5'},
		{'.','.','.','.','8','.','.','7','9'}
	};

	const vector<vector<char>> board2 =	{
		{'8','3','.','.','7','.','.','.','.'},
		{'6','.','.','1','9','5','.','.','.'},
		{'.','9','8','.','.','.','.','6','.'},
		{'8','.','.','.','6','.','.','.','3'},
		{'4','.','.','8','.','3','.','.','1'},
		{'7','.','.','.','2','.','.','.','6'},
		{'.','6','.','.','.','.','2','8','.'},
		{'.','.','.','4','1','9','.','.','5'},
		{'.','.','.','.','8','.','.','7','9'}
	};

	bool result = isValidSudoku(board);
	cout << "result: " << boolalpha << result << endl;
	bool result2 = isValidSudoku(board2);
	cout << "result: " << boolalpha << result2 << endl;

	return 0;
}

File: 36/valid_sudoku.js
"use strict";
exports.__esModule = true;
var createSetArray = function (length) { return Array(length).fill(0).map(function (_) { return new Set(); }); };
var createSubBoard = function (i, j) { return 3 * Math.floor(j / 3) + Math.floor(i / 3); };
function isValidSudoku(board) {
    var rows = createSetArray(board.length);
    var cols = createSetArray(board.length);
    var sub_boards = createSetArray(board.length);
    for (var i = 0; i < board.length; i++) {
        for (var j = 0; j < board[0].length; j++) {
            var cell = board[i][j];
            if (cell === ".")
                continue;
            var sub_board = createSubBoard(i, j);
            if (rows[i].has(cell) || cols[j].has(cell) || sub_boards[sub_board].has(cell)) {
                return false;
            }
            else {
                rows[i].add(cell);
                cols[j].add(cell);
                sub_boards[sub_board].add(cell);
            }
        }
    }
    return true;
}
function main() {
    var board = [
        ["5", "3", ".", ".", "7", ".", ".", ".", "."],
        ["6", ".", ".", "1", "9", "5", ".", ".", "."],
        [".", "9", "8", ".", ".", ".", ".", "6", "."],
        ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
        ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
        ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
        [".", "6", ".", ".", ".", ".", "2", "8", "."],
        [".", ".", ".", "4", "1", "9", ".", ".", "5"],
        [".", ".", ".", ".", "8", ".", ".", "7", "9"]
    ];
    var board2 = [
        ["8", "3", ".", ".", "7", ".", ".", ".", "."],
        ["6", ".", ".", "1", "9", "5", ".", ".", "."],
        [".", "9", "8", ".", ".", ".", ".", "6", "."],
        ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
        ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
        ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
        [".", "6", ".", ".", ".", ".", "2", "8", "."],
        [".", ".", ".", "4", "1", "9", ".", ".", "5"],
        [".", ".", ".", ".", "8", ".", ".", "7", "9"]
    ];
    var result = isValidSudoku(board);
    var result2 = isValidSudoku(board2);
    console.log(result);
    console.log(result2);
}
main();
